src/rule.cc

```diff
@@ -867,46 +882,101 @@ vsjson_test (bool verbose)
     printf (" * vsjson: skip\n");
 }
 
-void json_rule_test(const char *dir, const char *basename)
+void rule_test_json(const char *dir, const char *basename)
 {
+    assert(dir && basename);
+
     rule_t *self = rule_new ();
     assert (self);
-    char *rule_file = zsys_sprintf ("%s/rules/%s.rule", dir, basename);
-    char *json_file = zsys_sprintf ("%s/rules/%s.json", dir, basename);
-    assert (rule_file && json_file);
-    rule_load (self, rule_file);
-    zstr_free (&rule_file);
-    FILE *f;
-    char *stock_json;
-    assert (f = fopen (json_file, "r"));
-    assert (stock_json = (char *)calloc (1, 4096));
-    assert (fread (stock_json, 1, 4096, f));
-    fclose (f);
-    zstr_free (&json_file);
-    // test rule to json
-    char *json = rule_json (self);
-    // XXX: This is fragile, as we require the json to be bit-identical.
-    // If you get an error here, manually review the actual difference.
-    // In particular, the hash order is not stable
-    if (!streq (json, stock_json)) {
-        fprintf (stderr, "Generated json is different\nEXPECTED:\n%sGOT:\n%s",
-                stock_json, json);
-        abort ();
-    }
-    zstr_free (&stock_json);
+
+    {
+        char *rule_file = zsys_sprintf ("%s/%s.rule", dir, basename);
+        assert (rule_file);
+        int r = rule_load (self, rule_file);
+        assert(r == 0);
+        zstr_free (&rule_file);
+    }
+
+    char *json = NULL;
+    {
+        // read json related file
+        char *stock_json = NULL;
+        {
+            char *json_file = zsys_sprintf ("%s/%s.json", dir, basename);
+            assert (json_file);
+            FILE *f = fopen (json_file, "r");
+            assert (f);
+            stock_json = (char *)calloc (1, 4096);
+            assert (stock_json);
+            assert (fread (stock_json, 1, 4096, f));
```

Comment:
A not NULL-terminated C string will be created if the file is exactly 4096 bytes long.